ci/management/commands/generate_reports.py

```diff
@@ -405,3 +421,46 @@ def handle_enrollments(self, sheet, sbm_client, ids_client, start_date,
                 5: data[key]['optouts'],
                 6: data[key]['completed'],
             })
+
+    def handle_optouts_by_subscription(
+            self, sheet, sbm_client, ids_client, start_date, end_date):
+
+        sheet.set_header([
+            "Timestamp",
+            "Subscription Message Set",
+            "Receiver's Role",
+            "Reason",
+        ])
+
+        optouts = self.get_optouts(
+            ids_client, created_at__gte=start_date.isoformat(),
+            created_at__lte=end_date.isoformat())
+
+        for optout in optouts:
+            if 'identity' not in optout or not optout['identity']:
+                message_set = "Unknown"
+                receivers_role = "Unknown"
+            else:
+                identity = self.get_identity(ids_client, optout['identity'])
+                receivers_role = identity.get('data', {}).get(
+                    'role', 'Unknown')
+                # Get the last subscription before the optout that is inactive
+                subscriptions = list(self.get_subscriptions(
+                    sbm_client, identity=optout['identity'],
+                    created_to=optout['created_at'],
+                    active=False, completed=False))
+                subscriptions.sort(key=lambda s: s['created_at'], reverse=True)
+                try:
```

Comment:
I would prefer a if statement like `if subscriptions:`, but it works either way.